File: utils/SmartWallet.ts
import { Connection, PublicKey, Transaction, SystemProgram, LAMPORTS_PER_SOL } from '@solana/web3.js';
import { AnchorProvider, Program, BN } from '@project-serum/anchor';
import idl from '../public/idl2.json';

// Program ID from your smart wallet contract
const programId = new PublicKey('9RS7omQopJpsW3uYiCfxoTboEBtxo6a6o5GB5GCvmzUi');

// Vault program ID
const vaultProgramId = new PublicKey('GArNcH5X1sQka24mZvrGuA3QqDhvE9CBe35ZugwNevoH');

// Use the actual vault PDA from the vault contract
const vaultPda = new PublicKey('93XAG4BtLd4d6WQtuAfTAyg85yoXHMZiBrF9Aw8PcXvK');

/**
 * Get the wallet PDA for a given owner
 * @param owner Owner public key
 * @returns Wallet PDA
 */
export function getWalletPda(owner: PublicKey): PublicKey {
  const [walletPda] = PublicKey.findProgramAddressSync(
    [Buffer.from("wallet"), owner.toBuffer()],
    programId
  );
  return walletPda;
}

/**
 * Get the wallet account PDA for a given owner
 * @param owner Owner public key
 * @returns Wallet account PDA
 */
export function getWalletAccountPda(owner: PublicKey): PublicKey {
  const [walletAccountPda] = PublicKey.findProgramAddressSync(
    [Buffer.from("wallet_account"), owner.toBuffer()],
    programId
  );
  return walletAccountPda;
}

/**
 * Fund the wallet PDA with some SOL for rent exemption
 * @param wallet Phantom wallet instance
 * @param amount Amount in SOL to fund the wallet with
 * @returns Transaction signature
 */
export async function fundWalletPda(wallet: any, amount: number = 0.05): Promise<string> {
  if (!wallet?.publicKey) {
    throw new Error('Wallet not connected');
  }
  
  const connection = new Connection('https://api.devnet.solana.com', 'confirmed');
  const walletPda = getWalletPda(wallet.publicKey);
  
  // Create a transaction to transfer SOL from the user's wallet to the wallet PDA
  const transaction = new Transaction().add(
    SystemProgram.transfer({
      fromPubkey: wallet.publicKey,
      toPubkey: walletPda,
      lamports: amount * LAMPORTS_PER_SOL
    })
  );
  
  // Get a recent blockhash
  const { blockhash } = await connection.getRecentBlockhash();
  transaction.recentBlockhash = blockhash;
  transaction.feePayer = wallet.publicKey;
  
  // Sign the transaction
  const signed = await wallet.signTransaction(transaction);
  
  // Send the transaction
  const signature = await connection.sendRawTransaction(signed.serialize());
  
  // Wait for confirmation
  await connection.confirmTransaction(signature);
  
  return signature;
}

/**
 * Check if a wallet account is already initialized
 * @param wallet Phantom wallet instance
 * @returns boolean indicating if wallet is initialized
 */
export async function isWalletInitialized(wallet: any): Promise<boolean> {
  if (!wallet?.publicKey) {
    throw new Error('Wallet not connected');
  }
  
  const connection = new Connection('https://api.devnet.solana.com', 'confirmed');
  const walletAccountPda = getWalletAccountPda(wallet.publicKey);
  
  try {
    const accountInfo = await connection.getAccountInfo(walletAccountPda);
    return accountInfo !== null && accountInfo.owner.equals(programId);
  } catch (error) {
    console.error('Error checking wallet initialization:', error);
    return false;
  }
}

/**
 * Initialize a new smart wallet
 * @param wallet Phantom wallet instance
 * @returns Transaction signature
 */
export async function initializeWallet(wallet: any): Promise<string> {
  if (!wallet?.publicKey) {
    throw new Error('Wallet not connected');
  }
  
  // Check if wallet is already initialized
  const walletInitialized = await isWalletInitialized(wallet);
  if (walletInitialized) {
    throw new Error('Wallet is already initialized');
  }
  
  // First, make sure the wallet PDA has enough SOL for rent exemption
  try {
    const walletPda = getWalletPda(wallet.publicKey);
    const connection = new Connection('https://api.devnet.solana.com', 'confirmed');
    const balance = await connection.getBalance(walletPda);
    
    // If balance is too low, fund it
    if (balance < 0.02 * LAMPORTS_PER_SOL) {
      console.log('Funding wallet PDA first...');
      await fundWalletPda(wallet);
    }
  } catch (err) {
    console.error('Error checking or funding wallet PDA:', err);
  }
  
  const connection = new Connection('https://api.devnet.solana.com', 'confirmed');
  
  // Create provider with Phantom wallet
  const provider = new AnchorProvider(
    connection,
    {
      publicKey: wallet.publicKey,
      signTransaction: async (tx: Transaction) => {
        return await wallet.signTransaction(tx);
      },
      signAllTransactions: async (txs: Transaction[]) => {
        return await wallet.signAllTransactions(txs);
      }
    },
    { commitment: 'confirmed' }
  );
  
  // @ts-ignore - IDL type issues can be ignored for this example
  const program = new Program(idl, programId, provider);
  
  // Execute initialize wallet transaction
  const tx = await program.methods
    .initializeWallet()
    .accounts({
      owner: wallet.publicKey,
      wallet: getWalletAccountPda(wallet.publicKey),
      walletPda: getWalletPda(wallet.publicKey),
      systemProgram: PublicKey.default,
    })
    .rpc();
  
  return tx;
}

/**
 * Request funds from the vault and prepare for a transaction
 * @param amount Amount in SOL needed for the transaction
 * @param wallet Phantom wallet instance
 * @returns Transaction signature
 */
export async function requestFunds(amount: number, wallet: any): Promise<string> {
  if (!wallet?.publicKey) {
    throw new Error('Wallet not connected');
  }
  
  // Convert SOL to lamports
  const lamports = Math.floor(amount * LAMPORTS_PER_SOL);
  
  const connection = new Connection('https://api.devnet.solana.com', 'confirmed');
  
  // Create provider with Phantom wallet
  const provider = new AnchorProvider(
    connection,
    {
      publicKey: wallet.publicKey,
      signTransaction: async (tx: Transaction) => {
        return await wallet.signTransaction(tx);
      },
      signAllTransactions: async (txs: Transaction[]) => {
        return await wallet.signAllTransactions(txs);
      }
    },
    { commitment: 'confirmed', skipPreflight: false }
  );
  
  // Check if wallet account exists first
  const walletAccountPda = getWalletAccountPda(wallet.publicKey);
  const walletPda = getWalletPda(wallet.publicKey);
  
  // Verify account exists
  const accountInfo = await connection.getAccountInfo(walletAccountPda);
  if (!accountInfo) {
    throw new Error('Wallet account not initialized. Please initialize first.');
  }
  
  // @ts-ignore - IDL type issues can be ignored for this example
  const program = new Program(idl, programId, provider);
  
  try {
    // Execute request funds transaction
    const tx = await program.methods
      .executeTransaction(new BN(lamports))
      .accounts({
        wallet: walletAccountPda,
        walletPda: walletPda,
        vault: vaultPda,
        owner: wallet.publicKey,
        systemProgram: PublicKey.default,
      })
      .rpc();
    
    return tx;
  } catch (error: any) {
    console.error("Original error:", error);
    
    // If deserialization fails, try a more direct approach using low-level transaction
    if (error.toString().includes('AccountDidNotDeserialize') || 
        error.toString().includes('Failed to deserialize')) {
      
      console.log("Falling back to direct transaction...");
      
      // Build transaction manually
      const transaction = new Transaction();
      
      // Instruction data for executeTransaction - using a different approach for BigInt
      const instructionIndex = Buffer.from([1]); // instruction index 1 for executeTransaction
      const amountBuffer = Buffer.alloc(8);
      
      // Write u64 (8 bytes) in little-endian format manually
      for (let i = 0; i < 8; i++) {
        amountBuffer.writeUInt8(Number((BigInt(lamports) >> BigInt(8 * i)) & BigInt(255)), i);
      }
      
      // Combine the instruction index and amount
      const data = Buffer.concat([instructionIndex, amountBuffer]);
      
      // Add instruction to transaction
      transaction.add({
        keys: [
          { pubkey: walletAccountPda, isSigner: false, isWritable: false },
          { pubkey: walletPda, isSigner: false, isWritable: true },
          { pubkey: vaultPda, isSigner: false, isWritable: true },
          { pubkey: wallet.publicKey, isSigner: true, isWritable: false },
          { pubkey: PublicKey.default, isSigner: false, isWritable: false },
        ],
        programId: programId,
        data: data,
      });
      
      // Set recent blockhash and fee payer
      const { blockhash } = await connection.getRecentBlockhash();
      transaction.recentBlockhash = blockhash;
      transaction.feePayer = wallet.publicKey;
      
      // Sign transaction
      const signedTx = await wallet.signTransaction(transaction);
      
      // Send transaction
      const signature = await connection.sendRawTransaction(signedTx.serialize());
      await connection.confirmTransaction(signature);
      
      return signature;
    }
    
    // Re-throw if not a deserialization error
    throw error;
  }
}

/**
 * Return unused funds to the vault
 * @param amount Amount in SOL to return
 * @param wallet Phantom wallet instance
 * @returns Transaction signature
 */
export async function returnFunds(amount: number, wallet: any): Promise<string> {
  if (!wallet?.publicKey) {
    throw new Error('Wallet not connected');
  }
  
  // Convert SOL to lamports
  const lamports = Math.floor(amount * LAMPORTS_PER_SOL);
  
  const connection = new Connection('https://api.devnet.solana.com', 'confirmed');
  
  // Check if wallet account exists first
  const walletAccountPda = getWalletAccountPda(wallet.publicKey);
  const walletPda = getWalletPda(wallet.publicKey);
  
  // Verify account exists
  const accountInfo = await connection.getAccountInfo(walletAccountPda);
  if (!accountInfo) {
    throw new Error('Wallet account not initialized. Please initialize first.');
  }
  
  // Create provider with Phantom wallet
  const provider = new AnchorProvider(
    connection,
    {
      publicKey: wallet.publicKey,
      signTransaction: async (tx: Transaction) => {
        return await wallet.signTransaction(tx);
      },
      signAllTransactions: async (txs: Transaction[]) => {
        return await wallet.signAllTransactions(txs);
      }
    },
    { commitment: 'confirmed', skipPreflight: false }
  );
  
  // @ts-ignore - IDL type issues can be ignored for this example
  const program = new Program(idl, programId, provider);
  
  try {
    // Execute return funds transaction
    const tx = await program.methods
      .returnFunds(new BN(lamports))
      .accounts({
        wallet: walletAccountPda,
        walletPda: walletPda,
        vault: vaultPda,
        owner: wallet.publicKey,
        systemProgram: PublicKey.default,
      })
      .rpc();
    
    return tx;
  } catch (error: any) {
    console.error("Original error:", error);
    
    // If deserialization fails, try a more direct approach using low-level transaction
    if (error.toString().includes('AccountDidNotDeserialize') || 
        error.toString().includes('Failed to deserialize')) {
      
      console.log("Falling back to direct transaction...");
      
      // Build transaction manually
      const transaction = new Transaction();
      
      // Instruction data for returnFunds - using a different approach for BigInt
      const instructionIndex = Buffer.from([2]); // instruction index 2 for returnFunds
      const amountBuffer = Buffer.alloc(8);
      
      // Write u64 (8 bytes) in little-endian format manually
      for (let i = 0; i < 8; i++) {
        amountBuffer.writeUInt8(Number((BigInt(lamports) >> BigInt(8 * i)) & BigInt(255)), i);
      }
      
      // Combine the instruction index and amount
      const data = Buffer.concat([instructionIndex, amountBuffer]);
      
      // Add instruction to transaction
      transaction.add({
        keys: [
          { pubkey: walletAccountPda, isSigner: false, isWritable: false },
          { pubkey: walletPda, isSigner: false, isWritable: true },
          { pubkey: vaultPda, isSigner: false, isWritable: true },
          { pubkey: wallet.publicKey, isSigner: true, isWritable: false },
          { pubkey: PublicKey.default, isSigner: false, isWritable: false },
        ],
        programId: programId,
        data: data,
      });
      
      // Set recent blockhash and fee payer
      const { blockhash } = await connection.getRecentBlockhash();
      transaction.recentBlockhash = blockhash;
      transaction.feePayer = wallet.publicKey;
      
      // Sign transaction
      const signedTx = await wallet.signTransaction(transaction);
      
      // Send transaction
      const signature = await connection.sendRawTransaction(signedTx.serialize());
      await connection.confirmTransaction(signature);
      
      return signature;
    }
    
    // Re-throw if not a deserialization error
    throw error;
  }
}

/**
 * Get the wallet balance
 * @param wallet Phantom wallet instance
 * @returns Balance in SOL
 */
export async function getWalletBalance(wallet: any): Promise<number> {
  if (!wallet?.publicKey) {
    throw new Error('Wallet not connected');
  }
  
  const connection = new Connection('https://api.devnet.solana.com', 'confirmed');
  const walletPda = getWalletPda(wallet.publicKey);
  const balanceInLamports = await connection.getBalance(walletPda);
  return balanceInLamports / LAMPORTS_PER_SOL;
} 

File: src/smart_wallet.rs
use anchor_lang::prelude::*;
use anchor_lang::solana_program::{
    instruction::Instruction,
    program::invoke_signed,
};
use std::str::FromStr;

declare_id!("9RS7omQopJpsW3uYiCfxoTboEBtxo6a6o5GB5GCvmzUi"); // Update with your actual program ID

// Import the vault program ID
const VAULT_PROGRAM_ID: &str = "GArNcH5X1sQka24mZvrGuA3QqDhvE9CBe35ZugwNevoH";
// The actual vault PDA from the vault contract
const VAULT_PDA: &str = "93XAG4BtLd4d6WQtuAfTAyg85yoXHMZiBrF9Aw8PcXvK";

#[program]
pub mod smart_wallet {
    use super::*;

    // Initialize a new smart wallet for a user
    pub fn initialize_wallet(ctx: Context<InitializeWallet>) -> Result<()> {
        let wallet = &mut ctx.accounts.wallet;
        wallet.owner = ctx.accounts.owner.key();
        wallet.vault_program = Pubkey::from_str(VAULT_PROGRAM_ID).unwrap();
        wallet.vault_pda = Pubkey::from_str(VAULT_PDA).unwrap();

        msg!("Smart wallet initialized for owner: {}", wallet.owner);
        Ok(())
    }

    // Request funds from vault and execute transaction
    pub fn execute_transaction(
        ctx: Context<ExecuteTransaction>,
        amount: u64,
    ) -> Result<()> {
        let wallet = &ctx.accounts.wallet;
        
        // Verify owner
        require!(
            ctx.accounts.owner.key() == wallet.owner,
            WalletError::InvalidOwner
        );
        
        // Verify vault account
        require!(
            ctx.accounts.vault.key() == wallet.vault_pda,
            WalletError::InvalidVaultAccount
        );
        
        // Step 1: Request funds from the vault to the wallet PDA
        let withdraw_ix = anchor_lang::solana_program::instruction::Instruction {
            program_id: wallet.vault_program,
            accounts: vec![
                AccountMeta::new(ctx.accounts.vault.key(), false),
                AccountMeta::new(ctx.accounts.wallet_pda.key(), false),
                AccountMeta::new_readonly(ctx.accounts.owner.key(), true),
                AccountMeta::new_readonly(ctx.accounts.system_program.key(), false),
            ],
            data: anchor_lang::AnchorSerialize::try_to_vec(&WithdrawArgs {
                instruction: 2, // withdraw instruction index
                amount,
            })?,
        };
        
        // Calculate the wallet PDA seeds for signing
        let wallet_seeds = &[
            b"wallet",
            wallet.owner.as_ref(),
            &[ctx.bumps.wallet_pda],
        ];
        
        // Invoke the withdraw instruction
        invoke_signed(
            &withdraw_ix,
            &[
                ctx.accounts.vault.to_account_info(),
                ctx.accounts.wallet_pda.to_account_info(),
                ctx.accounts.owner.to_account_info(),
                ctx.accounts.system_program.to_account_info(),
            ],
            &[wallet_seeds],
        )?;
        
        msg!("Withdrew {} lamports from vault to wallet PDA", amount);
        
        // Now the wallet PDA has funds to pay for the transaction
        // The actual transaction execution will be handled by the client
        
        Ok(())
    }
    
    // Return unused funds back to the vault
    pub fn return_funds(ctx: Context<ReturnFunds>, amount: u64) -> Result<()> {
        let wallet = &ctx.accounts.wallet;
        
        // Verify owner
        require!(
            ctx.accounts.owner.key() == wallet.owner,
            WalletError::InvalidOwner
        );
        
        // Verify vault account
        require!(
            ctx.accounts.vault.key() == wallet.vault_pda,
            WalletError::InvalidVaultAccount
        );
        
        // Calculate the wallet PDA seeds for signing
        let wallet_seeds = &[
            b"wallet",
            wallet.owner.as_ref(),
            &[ctx.bumps.wallet_pda],
        ];
        
        // Create a system instruction to transfer funds back to the vault
        let transfer_ix = anchor_lang::solana_program::system_instruction::transfer(
            &ctx.accounts.wallet_pda.key(),
            &ctx.accounts.vault.key(),
            amount,
        );
        
        // Execute the transfer with PDA signature
        invoke_signed(
            &transfer_ix,
            &[
                ctx.accounts.wallet_pda.to_account_info(),
                ctx.accounts.vault.to_account_info(),
                ctx.accounts.system_program.to_account_info(),
            ],
            &[wallet_seeds],
        )?;
        
        msg!("Returned {} lamports to vault", amount);
        Ok(())
    }
}

#[derive(Accounts)]
pub struct InitializeWallet<'info> {
    #[account(mut)]
    pub owner: Signer<'info>,
    
    #[account(
        init,
        payer = owner,
        space = 8 + 32 + 32 + 32, // Discriminator + Owner + Vault Program ID + Vault PDA
        seeds = [b"wallet_account", owner.key().as_ref()],
        bump
    )]
    pub wallet: Account<'info, WalletAccount>,
    
    // PDA that will act as the wallet address
    /// CHECK: This is a PDA that will be used as a wallet
    #[account(
        seeds = [b"wallet", owner.key().as_ref()],
        bump,
    )]
    pub wallet_pda: AccountInfo<'info>,
    
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct ExecuteTransaction<'info> {
    #[account(
        has_one = owner @ WalletError::InvalidOwner,
    )]
    pub wallet: Account<'info, WalletAccount>,
    
    #[account(
        mut,
        seeds = [b"wallet", owner.key().as_ref()],
        bump,
    )]
    /// CHECK: This is the wallet PDA that will be used to sign transactions
    pub wallet_pda: AccountInfo<'info>,
    
    /// CHECK: This is verified in the vault program
    #[account(mut)]
    pub vault: AccountInfo<'info>,
    
    pub owner: Signer<'info>,
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct ReturnFunds<'info> {
    #[account()]
    pub wallet: Account<'info, WalletAccount>,
    
    #[account(
        mut,
        seeds = [b"wallet", owner.key().as_ref()],
        bump,
    )]
    /// CHECK: This is the wallet PDA that will be used to sign transactions
    pub wallet_pda: AccountInfo<'info>,
    
    /// CHECK: This is the vault account from the vault program
    #[account(mut)]
    pub vault: AccountInfo<'info>,
    
    pub owner: Signer<'info>,
    pub system_program: Program<'info, System>,
}

#[account]
pub struct WalletAccount {
    pub owner: Pubkey,         // The owner that can control this wallet
    pub vault_program: Pubkey, // The program ID of the vault
    pub vault_pda: Pubkey,     // The PDA of the vault account
}

// Define the withdraw arguments struct
#[derive(AnchorSerialize, AnchorDeserialize)]
pub struct WithdrawArgs {
    pub instruction: u8,
    pub amount: u64,
}

#[error_code]
pub enum WalletError {
    #[msg("Invalid owner for this wallet")]
    InvalidOwner,
    
    #[msg("Invalid vault account")]
    InvalidVaultAccount,
} 